File: dotnet/src/Pagou.Examples.Core/Webhooks/WebhookEvent.cs
using System.Text.Json.Nodes;

namespace Pagou.Examples.Core.Webhooks;

public enum WebhookFamily
{
    Transaction,
    Subscription,
    Transfer,
}

// Id is the top-level event id (the dedupe key). EventType is the concrete event
// name (e.g. transaction.paid). ResourceId is the affected resource's public id.
public sealed record WebhookEvent(
    string Id,
    WebhookFamily Family,
    string EventType,
    string ResourceId,
    JsonNode? Raw);

/// <summary>Either a parsed <see cref="WebhookEvent"/> or a typed ingestion failure.</summary>
public sealed record WebhookParseResult
{
    public WebhookEvent? Event { get; init; }

    /// <summary>"missing_event_id" | "unknown_envelope" when parsing failed.</summary>
    public string? Error { get; init; }

    public bool IsError => Error is not null;

    public static WebhookParseResult Ok(WebhookEvent evt) => new() { Event = evt };

    public static WebhookParseResult Fail(string error) => new() { Error = error };
}

File: php/webhooks/WebhookEvent.php
<?php

declare(strict_types=1);

namespace Pagou\Examples\Webhooks;

/** A parsed webhook, normalized across the three envelope families. */
final class WebhookEvent
{
    public function __construct(
        /** Top-level event id — the idempotency/dedupe key. */
        public readonly string $id,
        /** One of: transaction, subscription, transfer. */
        public readonly string $family,
        /** Concrete event name, e.g. `transaction.paid` or `payout.transferred`. */
        public readonly string $eventType,
        /** Public id of the affected resource, used to reconcile against the API. */
        public readonly string $resourceId,
        public readonly mixed $raw,
    ) {
    }
}
